File: C/sort/ShellSort.c
//C Program to sort an array using Shell Sort technique
#include <stdio.h>
//ShellSort Function
void ShellSort(int arr[], int num)
{
    int gap, i, j, tmp;
    //this for loop of gap is the heart of Shell sort
    //In each pass w check the valu of gap
    //if gap=1 then we stop this loop
    for (gap = num / 2; gap > 0; gap = gap / 2)
    {
        for (i = gap; i < num; i++)
        {
        	//loop continue for index lesser then gap
            for(j = i - gap; j >= 0; j = j - gap)
            {
            	//if lesser index value is small
            	//compared to bigger index
            	//then continue
                if (arr[j+gap] >= arr[j])
                    break;
                //else    
                // Swap values of bigger index with smaller one   
                else
                {
                    tmp = arr[j];
                    arr[j] = arr[j+gap];
                    arr[j+gap] = tmp;
                }
            }
        }
    }
}
//Display function to print the result
void display(int arr[],int n)
{
	int i;
printf("\n\nSorted Array in ascending order is : \n");
for(i=0; i < n; ++i)
printf("%5d",arr[i]);
}
//Main Function
int main()
{
    int i,k, num;
    //Enter the number of element array have
    printf("Enter total number of elements :: ");
    scanf("%d",&num);
    //store the element of array
    int arr[num];
    //Input values of array
    printf("Enter elements :: ");
    for(i = 0; i < num; i++)
    {
        scanf("%d",&arr[i]);
    }
    //Call shell sort function
    ShellSort(arr,num);
    //call display function
    display(arr,num);
    return 0;
}
//end of Main function

/*Sample Input Output
Enter total number of elements :: 10
Enter elements : 36 432 43 44 57 63  94 3 5 6

Sorted Array in ascending order is :
    3    5    6    36     43     44     57     63     94     432
*/

/*Complxity

Worst case time complexity is O(n2) 
Best case complexity is O(nlog(n)).
Space complexity is O(1).
*/

File: Python/sort/Shell_Sort.py
# Python program for implementation of Shell Sort
def shellSort(arr):

    # Start with a big gap, then reduce the gap
    size = len(arr)
    gap = size // 2

    # Do a gapped insertion sort for this gap size.
    while gap > 0:

        for i in range(gap, size):

            # add a[i] to the elements that have been gap sorted
            # save a[i] in temp and make a hole at position i
            temp = arr[i]

            # shift earlier gap-sorted elements up until the correct
            # location for a[i] is found
            j = i
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap

            # put temp (the original a[i]) in its correct location
            arr[j] = temp
        gap //= 2


arr = []
size = int(input('Enter size: '))
print ('Enter elements:')
for i in range(0, size):
    item = int(input())
    arr.append(item)

shellSort(arr)

print ('\nSorted Array:')
for i in range(size):
    print (arr[i])

'''
Sample I/O:
	Input: 
	Enter Size: 5
	Enter Elements: 
	5
	4
	3
	2
	1

	Output:
	Sorted Array:
	1
	2
	3
	4
	5

Time Complexity: O(n^2)
Space Complexity: O(1)
'''